File: Csharp/RT_CS/RT_CS/shapes/Rectangle.cs
using RT_CS.utility;
using RT_CS.Cameras;
using System;
using System.Collections.Generic;
using System.Text;

namespace RT_CS.shapes
{

    public class Rectangle : Triangle
    {
        private Vector3 vertexD,vdva,vdvc;

        public Rectangle(int id, int mid, Vector3 v1, Vector3 v2, Vector3 v3, Vector3 v4) : base(id, mid, v1, v2, v3)
        {
            vertexD = v4;
          
            vdva = v4 - v1;
            vdvc = v4 - v3;
            float vdvaMag = vdva.Magnitude();
            float vdvcMag = vdvc.Magnitude();
            area = vdvaMag * vdvcMag;
            normal = ((v3 - v4) * (v1 - v4)).Normalize();
            vdva /= vdvaMag;
            vdvc/=vdvcMag;
        }

        public override void Intersect(Ray r, IntersectionRecord intersection)
        {
            if (intersection.ExceptId == sid)
            {
                return;
            }
            IntersectionRecord inter = RayPlaneIntersection(r);
            if (inter.Hit)
            {
                float alpha = SimplerTriangleArea(vertexA - inter.Point,  vertexA - vertexD)/area;
                float beta = SimplerTriangleArea(vertexB - inter.Point,  vertexB - vertexA) /area;
                float gamma = SimplerTriangleArea( vertexC - inter.Point, vertexC - vertexB)/area;
                float phi = SimplerTriangleArea(vertexD - inter.Point, vertexD - vertexC)/area;
                float sum = alpha + beta + gamma + phi;
               
                if (Math.Abs(1 - sum) < BIGGER_EPSILON)
                {
                    if (inter.Distance < intersection.Distance)
                    {
                        intersection.Distance = inter.Distance;
                        intersection.Hit = true;
                        intersection.Point = inter.Point;
                        intersection.ShapeId = sid;
                    }
                }
            }
        }
    }


}

File: cpp/RT_CPP/RT_CPP/rectangle.cpp
#include <math.h>
#include "rectangle.h"
#include "operators.h"
using namespace shapes;

rectangle::rectangle(int id,int mid, vector3 a, vector3 b, vector3 c, vector3 d)
	: triangle(id,mid,a,b,c), vertexD(d)
{
    vavd = vertexD - vertexA;
    vavb = vertexB - vertexA;
    vavdMag = vavd.magnitude();
    vavbMag = vavb.magnitude();
	area = vavdMag * vavbMag;
    vavd = vavd / vavdMag;
    vavb = vavb / vavbMag;
}


vector2 rectangle::get_uv(vector3 point) const
{
    vector3 line = point - vertexA;
    float u = vector3::dot(line, vavb) / vavbMag;
    float v = vector3::dot(line, vavd) / vavdMag;
    return { u,v };
}

void rectangle::intersect(ray r, intersection_record& rec) const
{
    if (id == rec.except_id)
    {
        return;
    }
	intersection_record* ray_rec = ray_plane_intersection(r);
	if (ray_rec->hit)
	{
        float alpha = simpler_triangle_area(vertexD - vertexA, vertexD - ray_rec->point) / area;
        float beta = simpler_triangle_area(ray_rec->point - vertexA, ray_rec->point - vertexB) / area;
        float gamma = simpler_triangle_area(ray_rec->point - vertexB, ray_rec->point - vertexC) / area;
        float phi = simpler_triangle_area(vertexD - ray_rec->point, vertexD - vertexC) / area;
        float sum = alpha + beta + gamma + phi;

        if (fabsf(1 - sum) < BIGGER_EPSILON)
        {
            if (ray_rec->distance < rec.distance && ray_rec->distance > 0)
            {
                rec.distance = ray_rec->distance;
                rec.hit = true;
                rec.point = ray_rec->point;
                rec.shape_id = id;
            }
        }
	}
    delete ray_rec;
}
